### PR TITLE
[core][rdt] Implement pytorch storage block caching

### PR DESCRIPTION
**Pull request by Sparks0219**
A pytorch tensor has an underlying storage block that is managed by pytorch's memory allocator. For example, when you create a pytorch tensor and that tensor goes out of scope, the memory is not released to the OS. Pytorch caches it and manages it's own memory pool. 

You can see the memory address and size of the underlying pytorch tensor block by doing:
tensor.untyped_storage().data_ptr() 
tensor.untyped_storage().nbytes()
versus
tensor.data_ptr() 
tensor.nbytes()
which just gives you information about the view. A view is any tensor that shares the underlying storage block with another tensor, and can be a subset of that tensor. This is especially important for RDT since due to parallelism schemes (like tensor parallel) you could be be sending subsets of the weight tensor to different inference actors. This is often done along dimensions, hence you could have a 2d weight tensor where each row is a view, and send the row to a different actor.

What you would have to do right now to register all these views by ray.put-ing each view 1 by 1 where each view incurs a cost of calling the memory registration NIXL api. Now instead, you can ray.put the weight tensor once (so doing the memory registration on the entire weight tensor) and we will cache the memory registration. This also gives the advantage that RDT now supports overlapping views for the same storage block.  

Also did a minor refactor of the gc method to accept tensors instead of grabbing them from the gpu object store in the tensor transport class. Now nixl tensor transport is decoupled from the gpu object store.

### CODE REVIEW

File: python/ray/experimental/gpu_object_manager/nixl_tensor_transport.py [hunk 1]
```diff
@@ -391,17 +383,25 @@ def _put_meta(self, object_id: str, meta: NixlTransportMetadata):
...
-                if tensor.nbytes != self._tensor_desc_cache[key].nbytes:
-                    raise ValueError(
-                        "Tensors in an RDT object cannot partially overlap with each other."
-                    )
                 self._tensor_desc_cache[key].metadata_count += 1
             else:
-                reg_desc = self.get_nixl_agent().register_memory([tensor])
-                self._tensor_desc_cache[key] = TensorDesc(reg_desc, 1, tensor.nbytes)
+                mem_type = "cuda" if tensor.is_cuda else "cpu"
+                gpu_id = tensor.get_device()
```
[the remark this answers is not in the document]
Comment (Sparks0219) [reply python/ray/experimental/gpu_object_manager/nixl_tensor_transport.py]: Kick rocks Gemini, it returns -1 for CPU tensors https://docs.pytorch.org/docs/stable/generated/torch.Tensor.get_device.html

File: python/ray/experimental/gpu_object_manager/nixl_tensor_transport.py [hunk 2]
```diff
@@ -391,17 +383,25 @@ def _put_meta(self, object_id: str, meta: NixlTransportMetadata):
...
-                    )
                 self._tensor_desc_cache[key].metadata_count += 1
             else:
-                reg_desc = self.get_nixl_agent().register_memory([tensor])
-                self._tensor_desc_cache[key] = TensorDesc(reg_desc, 1, tensor.nbytes)
+                mem_type = "cuda" if tensor.is_cuda else "cpu"
+                gpu_id = max(tensor.get_device(), 0)
+                reg_desc = self.get_nixl_agent().register_memory(
+                    [
+                        (
```
Comment (stephanie-wang) [inline python/ray/experimental/gpu_object_manager/nixl_tensor_transport.py:375]: Can you add a comment explaining the args to this function?

File: python/ray/tests/gpu_objects/test_gpu_objects_nixl.py [hunk 3]
```diff
@@ -377,5 +377,29 @@ def test_nixl_agent_reuse_with_partial_tensors(ray_start_regular):
...
+        tensor = torch.tensor([[1, 1], [2, 2]], dtype=torch.float32).to("cuda")
+        views = [tensor[0], tensor[1]]
+        return ray.put(views, _tensor_transport="nixl")
+
+    def consume(self, ref):
+        return ray.get(ref)
+
+
+@pytest.mark.parametrize("ray_start_regular", [{"num_gpus": 2}], indirect=True)
+def test_storage_level_registration(ray_start_regular):
```
Comment (stephanie-wang) [inline python/ray/tests/gpu_objects/test_gpu_objects_nixl.py]: Hmm again we would need to check internal state here for the test to really test anything. It might be better to write a unit test here that directly calls the methods on the nixl tensor transport, instead of using integration tests through high-level ray calls like `ray.put`.

File: python/ray/experimental/gpu_object_manager/nixl_tensor_transport.py (hunk 2 above)
Comment (Sparks0219) [reply python/ray/experimental/gpu_object_manager/nixl_tensor_transport.py:375]: sounds good, done

File: python/ray/tests/gpu_objects/test_gpu_objects_nixl.py (hunk 3 above)
Comment (Sparks0219) [reply python/ray/tests/gpu_objects/test_gpu_objects_nixl.py]: makes sense, created a unit test

**Review by stephanie-wang — APPROVED** [review]
Looks good! Can you also add a test that we can now support overlapping tensors (can do in a follow-up too if you want to merge this first)?

**Comment by Sparks0219** [comment]
> Looks good! Can you also add a test that we can now support overlapping tensors (can do in a follow-up too if you want to merge this first)?

Thanks! I'll add it in this PR, and I'll modify the existing one to test partially overlapping tensors as well. I created a 2x3 tensor where view0 grabs the first 2 rows, and view1 grabs the bottom 2 rows so they share the middle. Hence we have a unit test making sure the ref counting is working properly, and I added an integration test to validate the ray.put layer just works as expected.

File: python/ray/tests/gpu_objects/test_gpu_objects_nixl.py [hunk 4]
```diff
@@ -377,5 +377,74 @@ def test_nixl_agent_reuse_with_partial_tensors(ray_start_regular):
...
+    view1 = tensor[1:2]
+    storage_key = tensor.untyped_storage().data_ptr()
+
+    mock_worker = MagicMock()
+    mock_worker.gpu_object_manager.gpu_object_store = gpu_object_store
+
+    # Simulate ray.put(view0)
+    obj_id1 = "test_obj_id_1"
+    gpu_object_store.add_object(obj_id1, [view0], is_primary=True)
+    with patch("ray._private.worker.global_worker", mock_worker):
```
Comment (dayshah) [inline python/ray/tests/gpu_objects/test_gpu_objects_nixl.py]: why do you need to mock out the worker in the tests?

Comment (Sparks0219) [reply python/ray/tests/gpu_objects/test_gpu_objects_nixl.py]: It was mainly because in the nixl tensor transport we call a gpu object store method and I needed direct access to it. However with https://github.com/ray-project/ray/pull/61122 I can prob remove the gpu object store usage completely

Comment (dayshah) [reply python/ray/tests/gpu_objects/test_gpu_objects_nixl.py]: can delete now

Comment (Sparks0219) [reply python/ray/tests/gpu_objects/test_gpu_objects_nixl.py]: hmmm still think I need it since the gc method in nixl_tensor_transport accesses the gpu object store

Comment (Sparks0219) [reply python/ray/tests/gpu_objects/test_gpu_objects_nixl.py]: Talked offline, better to have the gc method take in the tensors as thats part of a separate class

File: python/ray/tests/gpu_objects/test_gpu_objects_nixl.py [hunk 5]
```diff
@@ -377,5 +377,107 @@ def test_nixl_agent_reuse_with_partial_tensors(ray_start_regular):
...
+    actors = [OverlappingViewProducer.remote(), GPUTestActor.remote()]
+    src_actor, dst_actor = actors[0], actors[1]
+
+    refs = ray.get(src_actor.produce_overlapping_views.remote())
+    result = ray.get(dst_actor.consume_with_nixl.remote(refs))
+    assert result == 15
+
+
+@ray.remote(num_gpus=1, num_cpus=0, enable_tensor_transport=True)
+class WaitTensorFreedActor:
```
Comment (Sparks0219) [inline python/ray/tests/gpu_objects/test_gpu_objects_nixl.py:449]: This was added due to this thread: https://github.com/ray-project/ray/pull/61122#discussion_r2819839004

File: python/ray/experimental/gpu_object_manager/nixl_tensor_transport.py [hunk 6]
```diff
@@ -363,17 +354,31 @@ def _put_meta(self, object_id: str, meta: NixlTransportMetadata):
...
+                # Registering the full underlying pytorch storage object by constructing a memory region
+                # with the data pointer, size, GPU ID, and meta info. Doing the equivalent of what nixl does for pytorch tensors
+                # internally: https://github.com/ai-dynamo/nixl/blob/dd23ef01bd366aef89fa552f2b042f89a0b45fcb/src/api/python/_api.py#L1034
+                reg_desc = self.get_nixl_agent().register_memory(
+                    [
+                        (
+                            tensor.untyped_storage().data_ptr(),
+                            tensor.untyped_storage().nbytes(),
+                            gpu_id,
+                            "",
```
Comment (dayshah) [inline python/ray/experimental/gpu_object_manager/nixl_tensor_transport.py:379]: what's the empty string for?

**Review by dayshah — APPROVED** [review]
